### PR TITLE
Improve RocksJava Comparator

### PR DESCRIPTION
**Pull request by adamretter**
This is a redesign of the API for RocksJava comparators with the aim of improving performance. It also simplifies the class hierarchy.

**NOTE**: This breaks backwards compatibility for existing 3rd party Comparators implemented in Java... so we need to consider carefully which release branches this goes into.

Previously when implementing a comparator in Java the developer had a choice of subclassing either `DirectComparator` or `Comparator` which would use direct and non-direct byte-buffers resepectively (via `DirectSlice` and `Slice`).

In this redesign there we have eliminated the overhead of using the Java Slice classes, and just use `ByteBuffer`s. The `ComparatorOptions` supplied when constructing a Comparator allow you to choose between direct and non-direct byte buffers by setting `useDirect`.

In addition, the `ComparatorOptions` now allow you to choose whether a ByteBuffer is reused over multiple comparator calls, by setting `maxReusedBufferSize > 0`. When buffers are reused, ComparatorOptions provides a choice of mutex type by setting `useAdaptiveMutex`. 

---
[JMH benchmarks previously indicated](https://github.com/facebook/rocksdb/pull/6241#issue-356398306) that the difference between C++ and Java for implementing a comparator was ~7x slowdown in Java.

With these changes, when reusing buffers and guarding access to them via mutexes the slowdown is approximately the same. However, these changes offer a new facility to not reuse mutextes, which reduces the slowdown to ~5.5x in Java. We also offer a `thread_local` mechanism for reusing buffers, which reduces slowdown to ~5.2x in Java (closes https://github.com/facebook/rocksdb/pull/4425).

These changes also form a good base for further optimisation work such as further JNI lookup caching, and JNI critical.

---
These numbers were captured without jemalloc. With jemalloc, the performance improves for all tests, and the Java slowdown reduces to between 4.8x and 5.x.

```
ComparatorBenchmarks.put                                                native_bytewise  thrpt   25  124483.795 ± 2032.443  ops/s
ComparatorBenchmarks.put                                        native_reverse_bytewise  thrpt   25  114414.536 ± 3486.156  ops/s
ComparatorBenchmarks.put              java_bytewise_non-direct_reused-64_adaptive-mutex  thrpt   25   17228.250 ± 1288.546  ops/s
ComparatorBenchmarks.put          java_bytewise_non-direct_reused-64_non-adaptive-mutex  thrpt   25   16035.865 ± 1248.099  ops/s
ComparatorBenchmarks.put                java_bytewise_non-direct_reused-64_thread-local  thrpt   25   21571.500 ±  871.521  ops/s
ComparatorBenchmarks.put                  java_bytewise_direct_reused-64_adaptive-mutex  thrpt   25   23613.773 ± 8465.660  ops/s
ComparatorBenchmarks.put              java_bytewise_direct_reused-64_non-adaptive-mutex  thrpt   25   16768.172 ± 5618.489  ops/s
ComparatorBenchmarks.put                    java_bytewise_direct_reused-64_thread-local  thrpt   25   23921.164 ± 8734.742  ops/s
ComparatorBenchmarks.put                              java_bytewise_non-direct_no-reuse  thrpt   25   17899.684 ±  839.679  ops/s
ComparatorBenchmarks.put                                  java_bytewise_direct_no-reuse  thrpt   25   22148.316 ± 1215.527  ops/s
ComparatorBenchmarks.put      java_reverse_bytewise_non-direct_reused-64_adaptive-mutex  thrpt   25   11311.126 ±  820.602  ops/s
ComparatorBenchmarks.put  java_reverse_bytewise_non-direct_reused-64_non-adaptive-mutex  thrpt   25   11421.311 ±  807.210  ops/s
ComparatorBenchmarks.put        java_reverse_bytewise_non-direct_reused-64_thread-local  thrpt   25   11554.005 ±  960.556  ops/s
ComparatorBenchmarks.put          java_reverse_bytewise_direct_reused-64_adaptive-mutex  thrpt   25   22960.523 ± 1673.421  ops/s
ComparatorBenchmarks.put      java_reverse_bytewise_direct_reused-64_non-adaptive-mutex  thrpt   25   18293.317 ± 1434.601  ops/s
ComparatorBenchmarks.put            java_reverse_bytewise_direct_reused-64_thread-local  thrpt   25   24479.361 ± 2157.306  ops/s
ComparatorBenchmarks.put                      java_reverse_bytewise_non-direct_no-reuse  thrpt   25    7942.286 ±  626.170  ops/s
ComparatorBenchmarks.put                          java_reverse_bytewise_direct_no-reuse  thrpt   25   11781.955 ± 1019.843  ops/s
```

### CODE REVIEW

**Comment by adamretter** [comment]
@siying @pdillinger is it possible to get some review on this please?

**Comment by pdillinger** [comment]
> @pdillinger is it possible to get some review on this please?

Let's try to chat live so I can get more context on such a big change.

**Comment by adamretter** [comment]
@pdillinger okay, how do you want to do that? Slack?

File: java/src/main/java/org/rocksdb/AbstractComparator.java [hunk 1]
```diff
@@ -52,52 +58,115 @@ protected AbstractComparator(final ComparatorOptions copt) {
...
+  private int findShortSuccessorInternal(
+      final ByteBuffer key, final int keyLen) {
+    if (keyLen != -1) {
+      key.limit(keyLen);
+    }
+    findShortSuccessor(key);
+    return key.remaining();
+  }
+
+  public boolean usingDirectBuffers() {
```
Comment (pdillinger) [inline java/src/main/java/org/rocksdb/AbstractComparator.java]: Should be final?

File: java/src/main/java/org/rocksdb/AbstractComparator.java [hunk 2]
```diff
@@ -52,52 +58,115 @@ protected AbstractComparator(final ComparatorOptions copt) {
...
-   * @return a shorter key, or null
+   * @param key the key
    */
-  public String findShortSuccessor(final String key) {
-      return null;
+  public void findShortSuccessor(final ByteBuffer key) {
+    // no-op
   }
+
+  private int findShortSuccessorInternal(
```
Comment (pdillinger) [inline java/src/main/java/org/rocksdb/AbstractComparator.java]: Seems like these native-called private functions distract from the public API. Can they go in a package-protected class?

File: java/src/main/java/org/rocksdb/AbstractComparator.java [hunk 3]
```diff
@@ -52,52 +58,115 @@ protected AbstractComparator(final ComparatorOptions copt) {
...
+   * @param b buffer access to second key
    *
-   *  @return Should return either:
+   * @return Should return either:
    *    1) &lt; 0 if "a" &lt; "b"
    *    2) == 0 if "a" == "b"
    *    3) &gt; 0 if "a" &gt; "b"
    */
-  public abstract int compare(final T a, final T b);
+  public abstract int compare(final ByteBuffer a, final ByteBuffer b);
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply java/src/main/java/org/rocksdb/AbstractComparator.java]: In other words, if a comparator makes assumptions about the keys, such as their length, that could cause an exception or non-total order if violated, so the user is "living dangerously" and it is up to the user to ensure keys always satisfy those assumptions.

File: java/rocksjni/comparatorjnicallback.cc [hunk 4]
```diff
@@ -10,108 +10,270 @@
...
     releaseJniEnv(attached_thread);
     return 0;
   }
 
   jint result =
-    env->CallIntMethod(m_jcallback_obj, m_jCompareMethodId, m_jSliceA,
-      m_jSliceB);
+    env->CallIntMethod(m_jcallback_obj, m_jcompare_mid,
+      jcompare_buf_a, reuse_jbuf_a ? a.size() : -1,
+      jcompare_buf_b, reuse_jbuf_b ? b.size() : -1);
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply java/rocksjni/comparatorjnicallback.cc]: Oops, looks like I combined reuse_jbuf_a and reuse_jbuf_b in my head. Feel free to ignore

File: java/src/main/java/org/rocksdb/AbstractComparator.java (hunk 1 above)
Comment (adamretter) [reply java/src/main/java/org/rocksdb/AbstractComparator.java]: Yup, I will do that.

File: java/src/main/java/org/rocksdb/AbstractComparator.java (hunk 2 above)
Comment (adamretter) [reply java/src/main/java/org/rocksdb/AbstractComparator.java]: I will investigate moving these into a internal helper class with static methods as discussed.

File: java/src/main/java/org/rocksdb/AbstractComparator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdillinger) [reply java/src/main/java/org/rocksdb/AbstractComparator.java]: Pushed proposed new wording

**Comment by adamretter** [comment]
@pdillinger here are the benchmarks after internalising the Comparator JNI bridge stuff into package-private static methods:

```
ComparatorBenchmarks.put                          N/A                                                native_bytewise         N/A             N/A  thrpt   25   120953.855 ±  2183.177  ops/s
ComparatorBenchmarks.put                          N/A                                        native_reverse_bytewise         N/A             N/A  thrpt   25   110857.085 ±  5762.450  ops/s
ComparatorBenchmarks.put                          N/A              java_bytewise_non-direct_reused-64_adaptive-mutex         N/A             N/A  thrpt   25    17051.883 ±  1292.405  ops/s
ComparatorBenchmarks.put                          N/A          java_bytewise_non-direct_reused-64_non-adaptive-mutex         N/A             N/A  thrpt   25    16407.541 ±  1389.725  ops/s
ComparatorBenchmarks.put                          N/A                java_bytewise_non-direct_reused-64_thread-local         N/A             N/A  thrpt   25    22130.809 ±   857.235  ops/s
ComparatorBenchmarks.put                          N/A                  java_bytewise_direct_reused-64_adaptive-mutex         N/A             N/A  thrpt   25    22809.266 ±  8092.615  ops/s
ComparatorBenchmarks.put                          N/A              java_bytewise_direct_reused-64_non-adaptive-mutex         N/A             N/A  thrpt   25    17747.817 ±  5929.701  ops/s
ComparatorBenchmarks.put                          N/A                    java_bytewise_direct_reused-64_thread-local         N/A             N/A  thrpt   25    24107.863 ±  8746.386  ops/s
ComparatorBenchmarks.put                          N/A                              java_bytewise_non-direct_no-reuse         N/A             N/A  thrpt   25    17916.958 ±   816.778  ops/s
ComparatorBenchmarks.put                          N/A                                  java_bytewise_direct_no-reuse         N/A             N/A  thrpt   25    21966.780 ±  1374.856  ops/s
ComparatorBenchmarks.put                          N/A      java_reverse_bytewise_non-direct_reused-64_adaptive-mutex         N/A             N/A  thrpt   25    11672.043 ±   862.289  ops/s
ComparatorBenchmarks.put                          N/A  java_reverse_bytewise_non-direct_reused-64_non-adaptive-mutex         N/A             N/A  thrpt   25    11483.692 ±   893.066  ops/s
ComparatorBenchmarks.put                          N/A        java_reverse_bytewise_non-direct_reused-64_thread-local         N/A             N/A  thrpt   25    11562.274 ±   753.178  ops/s
ComparatorBenchmarks.put                          N/A          java_reverse_bytewise_direct_reused-64_adaptive-mutex         N/A             N/A  thrpt   25    22147.261 ±  1614.078  ops/s
ComparatorBenchmarks.put                          N/A      java_reverse_bytewise_direct_reused-64_non-adaptive-mutex         N/A             N/A  thrpt   25    18161.578 ±  1751.115  ops/s
ComparatorBenchmarks.put                          N/A            java_reverse_bytewise_direct_reused-64_thread-local         N/A             N/A  thrpt   25    24528.654 ±  2344.716  ops/s
ComparatorBenchmarks.put                          N/A                      java_reverse_bytewise_non-direct_no-reuse         N/A             N/A  thrpt   25     7650.531 ±   674.421  ops/s
ComparatorBenchmarks.put                          N/A                          java_reverse_bytewise_direct_no-reuse         N/A             N/A  thrpt   25    12172.327 ±   787.346  ops/s
```

**Comment by adamretter** [comment]
@pdillinger So I think we are good to go :-)

File: java/rocksjni/comparatorjnicallback.cc [hunk 5]
```diff
@@ -202,139 +429,214 @@ void BaseComparatorJniCallback::FindShortSuccessor(
...
+    // exception occurred
+    return nullptr;
   }
+  return jbuf;
+}
 
-  releaseJniEnv(attached_thread);
+void ComparatorJniCallback::DeleteBuffer(JNIEnv* env, jobject jbuffer) const {
+  // if direct then free delete the underlying array too
+  // if (m_options->direct_buffer) {
```
Comment (pdillinger) [inline java/rocksjni/comparatorjnicallback.cc]: What's the deal with this commented out code?

File: java/src/main/java/org/rocksdb/util/IntComparator.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+   * @param b 4-bytes representing an integer key
+   *
+   * @return negative if a &lt; b, 0 if a == b, positive otherwise
+   */
+  private final int compareIntKeys(final ByteBuffer a, final ByteBuffer b) {
+    final int iA = a.getInt();
+    final int iB = b.getInt();
+
+    // protect against int key calculation overflow
+    final double diff = (double)iA - iB;
```
[the remark this answers is not in the document]
Comment (adamretter) [reply java/src/main/java/org/rocksdb/util/IntComparator.java]: Sure will do

File: java/rocksjni/comparatorjnicallback.cc (hunk 5 above)
Comment (adamretter) [reply java/rocksjni/comparatorjnicallback.cc]: Its not needed I will remove it

**Comment by adamretter** [comment]
@pdillinger I have addressed your remaining issues. I think it's now ready to go

**Pull request closed without merging**